proposals/0293-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -514,6 +515,48 @@ struct TextEditor {
 }
 ```
 
+### Extending property wrappers to patterns
+
+Property-wrapper backing-storage initialization in the the pattern of a closure argument was supported in first revision. Building on this syntax, the core team suggested the extension of property-wrapper application to places where patterns exist. Of course, the design has been amended so as to preserve the expectation that the backing storage be private; extending property wrappers to patterns, though, is still a viable future direction. 
+
+Enabling the application of property wrappers where patterns are available is quite straightforward and could be introduced in as part of two separate features. The first could be to enable utility wrappers –– such as `@Asserted` –– in patterns. The second could be enabling projected-value initialization, which would facilitate intuitive and effortless access to property wrappers in native language constructs, as shown below:
```

Comment:
I don't understand the "two separate features" approach. For this line of code:

```swift
let (@Traceable userRatings, ... ) = ...
```

What type is the user providing on the right side? Is this initializing a new property wrapper, or "binding" `userRatings` to the wrapped value of an existing property wrapper? I thought the goal of property wrappers in patterns was the latter.

Reply:
Sorry for taking so long to reply!

`What type is the user providing on the right side?`

The user is providing the wrapped value's type; I'll clarify that.


` Is this initializing a new property wrapper, or "binding" userRatings to the wrapped value of an existing property wrapper?`

That's what I meant by two features: the first feature is initializing a new property wrapper that wraps a given value extracted by the pattern; the second is binding the wrapped declaration to an existing property wrapper. 

`I thought the goal of property wrappers in patterns was the latter.`

I think it could be both; being able to apply a wrapper to a value extracted by a pattern (i.e. a value extracted by the pattern found in `guard let`) would too be immensely useful. For instance, application of a wrapper to the entire –– "non-extracted" –– value is sometimes pointless (it does not make sense to apply `@Lowercased` to a `String?`-bound declaration). 

Do you think the feature should be separated into two different future directions?

Reply:
It's pretty clear that this future direction would involve a lot of design work. I think trying to get into that design here is distracting, so I think we should simplify the section and just say that it was suggested as a design for closures, we chose not to go with that design because it makes more sense for closures to behave the same way as functions, but property wrappers in patterns is still something that could be pursued in the future

Reply:
Does this look better?

----

Property-wrapper backing-storage initialization in the the pattern of a closure argument was supported in first revision. Building on this syntax, the core team suggested the extension of property-wrapper application to places where patterns exist. Of course, the design has been amended so as to preserve the expectation that the backing storage be private; extending property wrappers to patterns, though, is still a viable future direction.

Enabling the application of property wrappers where patterns are available is quite straightforward. It would simply require enabling projected-value initialization, which would facilitate intuitive and effortless access to property wrappers in native language constructs, as shown below:

```swift
enum Review {
  case revised(History<String>), original(String)
  
  init(fromUser username: String) { ... }
}

switch Review(fromUser: "swiftUser5") {
case .revised(@Traceable let $reviewText):
  ...
case .original(let originalReview):
  ...
}
```